doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,571 @@
+    "frame_code": {
+        "global_code": str,
+        "local_code": str,
+        "function_signatures": Dict[str, str],
+        "header_includes": List[str],
+    },
+
+    # From TargetCodeGenerationPass
+    "target_code": {
+        "code_objects": List[CodeObject],
+        "target_specific_code": Dict[str, str],
+        "compilation_flags": Dict[str, List[str]],
+    },
+}
+```
+
+### Information Reuse Strategies
+
+1. **Caching Analysis Results**: Store expensive analysis results (allocation lifetime, control flow) to avoid recomputation
+2. **Incremental Updates**: Allow passes to update only changed portions of the SDFG
+3. **Shared Data Structures**: Use common data structures across passes for consistency
+4. **Lazy Evaluation**: Compute expensive analyses only when needed by downstream passes
+
+## Organizational Structure
+
+### Current Structure
+```
+dace/codegen/
+├── __init__.py
+├── codegen.py              # Main entry point
+├── targets/                # Target-specific generators
+├── instrumentation/        # Profiling support
+├── tools/                  # Utilities
+├── dispatcher.py           # Target routing
+├── control_flow.py         # Control flow analysis
+├── compiler.py             # CMake integration
+├── codeobject.py          # Code file representation
+├── compiled_sdfg.py       # Runtime interface
+└── ...
+```
+
+### Proposed Structure
+```
+dace/codegen/
+├── __init__.py
+├── codegen.py              # Simplified entry point
+├── compiler/               # Compilation and build system
+│   ├── __init__.py
+│   ├── cmake.py            # CMake integration (from compiler.py)
+│   ├── direct.py           # Direct compiler calls
+│   ├── build_system.py     # Abstract build system interface
+│   └── environments.py     # Environment-specific build logic
+├── passes/                 # Code generation passes
+│   ├── __init__.py
+│   ├── analysis/           # Analysis passes
+│   │   ├── __init__.py
+│   │   ├── type_inference.py
+│   │   ├── metadata_collection.py
+│   │   ├── allocation_analysis.py
+│   │   ├── control_flow_analysis.py
+│   │   └── target_analysis.py
+│   ├── transformation/     # Transformation passes
+│   │   ├── __init__.py
+│   │   ├── copy_to_map.py
+│   │   ├── stream_assignment.py
+│   │   ├── tasklet_lowering.py
+│   │   └── memlet_lowering.py
+│   ├── codegen/           # Code generation passes
+│   │   ├── __init__.py
+│   │   ├── state_struct.py
+│   │   ├── allocation_code.py
+│   │   ├── frame_code.py
+│   │   ├── target_code.py
+│   │   └── header_generation.py
+│   ├── file_generation/   # File organization passes
+│   │   ├── __init__.py
+│   │   ├── sdfg_splitting.py
+│   │   └── code_objects.py
+│   └── pipelines.py       # Pre-built pipelines
+├── targets/               # Target-specific generators (simplified)
+│   ├── __init__.py
+│   ├── base.py            # Base target interface (from target.py)
+│   ├── openmp.py          # OpenMP backend (split from cpu.py)
```

Comment:
It looks like we are adding an OpenMP-specific target, so should we have a corresponding OpenMPCodegenPipeline?

Reply:
Since the target-specific pipelines were too detailed for a design doc, this is no longer necessary to describe. We do want to enforce creating openmp.py though because of the cross-target entanglement